perun/view/scatter_plot/factory.py

```diff
@@ -0,0 +1,156 @@
+""" Module with graphs creation and configuration functions. """
+
+from operator import itemgetter
+from collections import defaultdict
+
+import perun.profile.query as query
+import perun.profile.converters as converters
+import perun.utils.bokeh_helpers as bokeh_helpers
+
+import demandimport
+with demandimport.enabled():
+    import bkcharts as charts
+    import bokeh.palettes as palettes
+
+__author__ = 'Jiri Pavela'
+
+
+def slice_resources_by_uid(resources, models, uids):
+    """ Splits the resource tables and models into slices by the unique uids found in the resources.
+
+    Arguments:
+        resources(pandas.DataFrame): the data table from resources
+        models(list of dict): the list of models from profile
+        uids(list): the list of unique uids from profile
+
+    Returns:
+        generator: resources and models slices of unique uid as pair
+            data_slice(pandas.DataFrame)
+            uid_models(list)
+    """
+    for uid in uids:
+        # Slice only the plotted uid from the data table
+        uid_slice = resources[resources.uid == uid]
+        if uid_slice.size == 0 or uid_slice.shape[0] <= 1:
+            # plotting one point does not work (it has no real usage anyway), fix later
+            continue
+        # Filter models for the given uid
+        uid_models = list(filter(lambda m, u=uid: m['uid'] in u, models))
+        yield uid_slice, uid_models
+
+
+def slice_models_by_interval(models):
+    """ Splits the models list into slices with different x axis intervals.
+
+    Arguments:
+        models(list of dict): the list of models to split
+
+    Returns:
+        generator: stream of models slices (list)
+    """
+    # Sort the models by intervals first, to yield them in order
+    models = sorted(models, key=itemgetter('x_interval_start', 'x_interval_end'))
+    # Separate the models into groups according to intervals
+    intervals = defaultdict(list)
+    for model in models:
+        intervals[(model['x_interval_start'], model['x_interval_end'])].append(model)
+    # Yield the list of models with the same interval
+    for interval_models in intervals.items():
+        yield interval_models[1]
+
+
+def generate_plot_data_slices(profile):
+    """ Generates data slices for plotting resources and models. The resources are split by unique
+        uids, models are sliced into parts by uid and interval.
+
+    Arguments:
+        profile(dict): loaded perun profile
+
+    Returns:
+        generator: generator: resources and models slices of unique uid as pair
+            data_slice(pandas.DataFrame)
+            uid_models(list)
+    """
+    # Get resources for scatter plot points and models for curves
+    resource_table = converters.resources_to_pandas_dataframe(profile)
+    models = [m[1] for m in list(query.all_models_of(profile))]
```

Comment:
Mm, maybe list(map(itemgetter(1), querry.all_models_of(profile)))*?

(*) Not sure if this is syntactically correct though.

Reply:
Changed.